internal/graph/limited_checker.go

```diff
@@ -0,0 +1,60 @@
+package graph
+
+import (
+	"context"
+	"fmt"
+)
+
+type LimitedLocalChecker struct {
+	resolver CheckResolver
+	limiter  chan struct{}
+}
+
+type LimitedLocalCheckerOpt func(l *LimitedLocalChecker)
+
+func WithCheckResolver(resolver CheckResolver) LimitedLocalCheckerOpt {
+	return func(l *LimitedLocalChecker) {
+		l.resolver = resolver
+	}
+}
+
+func WithConcurrencyLimit(limit uint32) LimitedLocalCheckerOpt {
+	return func(l *LimitedLocalChecker) {
+		l.limiter = make(chan struct{}, limit)
+	}
+}
+
+func MustNewLimitedLocalChecker(opts ...LimitedLocalCheckerOpt) *LimitedLocalChecker {
+	checker, err := NewLimitedLocalChecker(opts...)
+	if err != nil {
+		panic(err)
+	}
+
+	return checker
+}
+
+func NewLimitedLocalChecker(opts ...LimitedLocalCheckerOpt) (*LimitedLocalChecker, error) {
+
+	checker := &LimitedLocalChecker{
+		limiter: make(chan struct{}, 50),
```

Comment:
Should make this a const `defaultConcurrencyLimit` or something.

Reply:
The lack of channel closure _should_ be fine, because the other branches are cancelled and so the Go garbage collector should be able to reclaim the unused channel. You don't always have to explicitly close a channel in Go. The Go garbage collector can use escape analysis to determine the channel is no longer in use.

> It is only necessary to close a channel if the receiver is looking for a close. Closing the channel is a control signal on the channel indicating that no more data follows.

Closure is a signaling mechanism mostly. Otherwise Go GC can use escape analysis and collect it automatically. If, however, you have goroutines lying around that still reference the channel, then that is considered a memory leak (both because of the dangling goroutines and the channel that is still being referenced by said goroutines).

Reply:
Agree on making this value configurable as depending on the database configuration, this value may be different.